src/pyhf/infer/calculators.py

```diff
@@ -168,3 +174,75 @@ def _false_case():
                 (sqrtqmu_v < self.sqrtqmuA_v), _true_case, _false_case
             )
         return teststat
+
+
+class EmpiricalDistribution(object):
+    def __init__(self, samples):
+        self.samples = samples.ravel()
+
+    def pvalue(self, value):
+        tensorlib, _ = get_backend()
+        return (
+            tensorlib.where(self.samples >= value, 1, 0).sum()
+            / tensorlib.shape(self.samples)[0]
+        )
+
+    def expected_value(self, nsigma):
+        tensorlib, _ = get_backend()
+        import numpy as np
+
+        # TODO: tensorlib.percentile function
+        return np.percentile(self.samples, (tensorlib.normal_cdf(nsigma)) * 100)
+
+
+class ToyCalculator(object):
+    def __init__(
+        self, data, pdf, init_pars=None, par_bounds=None, qtilde=False, ntoys=2000
+    ):
+        self.ntoys = ntoys
+        self.data = data
+        self.pdf = pdf
+        self.init_pars = init_pars or pdf.config.suggested_init()
+        self.par_bounds = par_bounds or pdf.config.suggested_bounds()
+
+    def distributions(self, poi_test):
+        tensorlib, _ = get_backend()
+        sample_shape = (self.ntoys,)
+
+        signal_pars = self.pdf.config.suggested_init()
+        signal_pars[self.pdf.config.poi_index] = poi_test
+        signal_pdf = self.pdf.make_pdf(tensorlib.astensor(signal_pars))
+        signal_sample = signal_pdf.sample(sample_shape)
+
+        bkg_pars = self.pdf.config.suggested_init()
+        bkg_pars[self.pdf.config.poi_index] = 0.0
+        bkg_pdf = self.pdf.make_pdf(tensorlib.astensor(bkg_pars))
+        bkg_sample = bkg_pdf.sample(sample_shape)
+
+        print('signal!')
+        signal_qtilde = []
+        for i, sample in enumerate(signal_sample):
+            if i % 100 == 0:
+                print(i)
+            signal_qtilde.append(
+                qmu(poi_test, sample, self.pdf, signal_pars, self.par_bounds)
+            )
+        signal_qtilde = tensorlib.astensor(signal_qtilde)
+
```

Comment:
the progress reporting here should be improved.. alsoo progress bars? something else? I want to avoid cluttering up the stdout but even w/ pyhf toys can be slow

also do we want parallelism a la joblib?

Reply:
> progress bars?

Would be nice. :) I don't think there are too many times that I've seen a progress bar and thought to myself "oh I hate that".

> do we want parallelism a la joblib?

Parallelism sounds generally preferable.

Reply:
let's pull this feature into a new issue. out of scope for this PR i think.

Reply:
#807 is filed.

Reply:
Progress bars via tqdm added in this PR.